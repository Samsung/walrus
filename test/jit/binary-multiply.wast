(module

(func (export "test1") (param i64) (result i64 i64 i64)
  i64.const 10000000
  local.get 0
  i64.mul

  i64.const -10000000
  local.get 0
  i64.mul

  i64.const -10000000
  local.set 0

  local.get 0
  local.get 0
  i64.mul

  (; 100000000000000, 18446644073709551616, 100000000000000 ;)
)

(func (export "test2") (param i32) (result i32 i32 i32)
  i32.const 10000
  local.get 0
  i32.mul

  i32.const -10000
  local.get 0
  i32.mul

  i32.const -10000
  local.set 0

  local.get 0
  local.get 0
  i32.mul

  (; 100000000, 4194967296, 100000000 ;)
)

(func (export "test3") (param i64) (result i64 i64)
  i64.const 0x3bc5600000000
  local.get 0
  i64.mul

  i64.const 0x2d961e00000001
  local.get 0
  i64.mul
)

(func (export "test4") (param i64) (result i64 i64)
  i64.const 0x572c3
  local.get 0
  i64.mul

  i64.const 0x160452d48
  local.get 0
  i64.mul
)

(func (export "test5") (param i32 i64) (result i64)
  local.get 0
  i32.const 4
  i32.add

  local.get 1
  i64.const 0x1234
  i64.add
  local.set 1

  local.set 0

  local.get 1
  local.get 1
  i64.mul
)

(func (export "test6") (param i32 i64) (result i64)
  local.get 0
  i32.const 4
  i32.add

  local.get 1
  i64.const 0x4321
  i64.add
  local.set 1

  local.set 0

  local.get 1
  i64.const 0x48b71cd0253
  i64.mul
)
)

(assert_return (invoke "test1" (i64.const 10000000)) (i64.const 100000000000000) (i64.const 18446644073709551616) (i64.const 100000000000000))
(assert_return (invoke "test2" (i32.const 10000)) (i32.const 100000000) (i32.const 4194967296) (i32.const 100000000))
(assert_return (invoke "test3" (i64.const 0x34b74b74b74b74b7)) (i64.const 0x5b34997a00000000) (i64.const 0x6a1d32e6b74b74b7))
(assert_return (invoke "test4" (i64.const 0x580f80f80f80f80f)) (i64.const 0x8016016015fea16d) (i64.const 0x5635635583eb6738))
(assert_return (invoke "test5" (i32.const 0) (i64.const 0xc48a2406a2e5)) (i64.const 0x219b6a1005485c71))
(assert_return (invoke "test6" (i32.const 0) (i64.const 0xd4b913dfe24a)) (i64.const 0x5733424463a5f7b1))
