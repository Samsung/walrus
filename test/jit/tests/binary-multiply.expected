[[[[[[[  Function   0  ]]]]]]]
Local 0: mapped to: 0 range [1 - 11]
Number of locals reduced from 1 to 1
0: Opcode: i64.const
  Result[0]: i64
1: Opcode: local.set
  Index: 0
  Param[0]: ref:0(0)
2: Opcode: i64.const
  Result[0]: i64
3: Opcode: local.get
  Index: 0
  Result[0]: i64
4: Opcode: i64.mul
  Param[0]: ref:2(0)
  Param[1]: ref:3(0)
  Result[0]: i64
5: Opcode: i64.const
  Result[0]: i64
6: Opcode: local.get
  Index: 0
  Result[0]: i64
7: Opcode: i64.mul
  Param[0]: ref:5(0)
  Param[1]: ref:6(0)
  Result[0]: i64
8: Opcode: i64.const
  Result[0]: i64
9: Opcode: local.set
  Index: 0
  Param[0]: ref:8(0)
10: Opcode: local.get
  Index: 0
  Result[0]: i64
11: Opcode: local.get
  Index: 0
  Result[0]: i64
12: Opcode: i64.mul
  Param[0]: ref:10(0)
  Param[1]: ref:11(0)
  Result[0]: i64
13: Opcode: return
  Param[0]: ref:4(0)
  Param[1]: ref:7(0)
  Param[2]: ref:12(0)
------------------------------
Frame size: 0, Arguments size: 32 (total: 32)
0: Opcode: i64.const
  Result[0]: i64 unused
1: Opcode: local.set
  Addr: sp[24]
  Param[0]: i64 const:0
2: Opcode: i64.const
  Result[0]: i64 unused
3: Opcode: local.get
  Addr: sp[24]
  Result[0]: i64 unused
4: Opcode: i64.mul
  Param[0]: i64 const:2
  Param[1]: i64 sp[24]
  Result[0]: i64 sp[0]
5: Opcode: i64.const
  Result[0]: i64 unused
6: Opcode: local.get
  Addr: sp[24]
  Result[0]: i64 unused
7: Opcode: i64.mul
  Param[0]: i64 const:5
  Param[1]: i64 sp[24]
  Result[0]: i64 sp[8]
8: Opcode: i64.const
  Result[0]: i64 unused
9: Opcode: local.set
  Addr: sp[24]
  Param[0]: i64 const:8
10: Opcode: local.get
  Addr: sp[24]
  Result[0]: i64 unused
11: Opcode: local.get
  Addr: sp[24]
  Result[0]: i64 unused
12: Opcode: i64.mul
  Param[0]: i64 sp[24]
  Param[1]: i64 sp[24]
  Result[0]: i64 sp[16]
13: Opcode: return
  Param[0]: i64 sp[0]
  Param[1]: i64 sp[8]
  Param[2]: i64 sp[16]

[[[[[[[  Function   1  ]]]]]]]
Local 0: mapped to: 0 range [1 - 11]
Number of locals reduced from 1 to 1
0: Opcode: i32.const
  Result[0]: i32
1: Opcode: local.set
  Index: 0
  Param[0]: ref:0(0)
2: Opcode: i32.const
  Result[0]: i32
3: Opcode: local.get
  Index: 0
  Result[0]: i32
4: Opcode: i32.mul
  Param[0]: ref:2(0)
  Param[1]: ref:3(0)
  Result[0]: i32
5: Opcode: i32.const
  Result[0]: i32
6: Opcode: local.get
  Index: 0
  Result[0]: i32
7: Opcode: i32.mul
  Param[0]: ref:5(0)
  Param[1]: ref:6(0)
  Result[0]: i32
8: Opcode: i32.const
  Result[0]: i32
9: Opcode: local.set
  Index: 0
  Param[0]: ref:8(0)
10: Opcode: local.get
  Index: 0
  Result[0]: i32
11: Opcode: local.get
  Index: 0
  Result[0]: i32
12: Opcode: i32.mul
  Param[0]: ref:10(0)
  Param[1]: ref:11(0)
  Result[0]: i32
13: Opcode: return
  Param[0]: ref:4(0)
  Param[1]: ref:7(0)
  Param[2]: ref:12(0)
------------------------------
Frame size: 0, Arguments size: 16 (total: 16)
0: Opcode: i32.const
  Result[0]: i32 unused
1: Opcode: local.set
  Addr: sp[12]
  Param[0]: i32 const:0
2: Opcode: i32.const
  Result[0]: i32 unused
3: Opcode: local.get
  Addr: sp[12]
  Result[0]: i32 unused
4: Opcode: i32.mul
  Param[0]: i32 const:2
  Param[1]: i32 sp[12]
  Result[0]: i32 sp[0]
5: Opcode: i32.const
  Result[0]: i32 unused
6: Opcode: local.get
  Addr: sp[12]
  Result[0]: i32 unused
7: Opcode: i32.mul
  Param[0]: i32 const:5
  Param[1]: i32 sp[12]
  Result[0]: i32 sp[4]
8: Opcode: i32.const
  Result[0]: i32 unused
9: Opcode: local.set
  Addr: sp[12]
  Param[0]: i32 const:8
10: Opcode: local.get
  Addr: sp[12]
  Result[0]: i32 unused
11: Opcode: local.get
  Addr: sp[12]
  Result[0]: i32 unused
12: Opcode: i32.mul
  Param[0]: i32 sp[12]
  Param[1]: i32 sp[12]
  Result[0]: i32 sp[8]
13: Opcode: return
  Param[0]: i32 sp[0]
  Param[1]: i32 sp[4]
  Param[2]: i32 sp[8]

test1() => i64:100000000000000, i64:18446644073709551616, i64:100000000000000
test2() => i32:100000000, i32:4194967296, i32:100000000
